### PR TITLE
Enable transpose scheduler

### PR DESCRIPTION
**Pull request by zasdfgbnm**
# transpose benchmarks (RTX 3090)

Command:
```
./build/bin/nvfuser_bench --benchmark_out=filename.csv --benchmark_out_format=csv "--benchmark_filter=.*Transpose.*
```
Raw data (before): https://github.com/zasdfgbnm/things/blob/master/2022/nvfuser-benchmarks/transpose-pointwise.csv
Raw data (after): https://github.com/zasdfgbnm/things/blob/master/2022/nvfuser-benchmarks/transpose-transpose.csv

Screenshot_20220911_000425

| Geomean | 1.14  |
|---------|-------|
| max     | 12.08 |
| top 1%  | 9.26  |
| top 10% | 1.43  |
| top 50% | 1.00  |
| top 90% | 0.97  |
| top 99% | 0.96  |
| min     | 0.95  |

# Benchmarks of the entire suite (RTX 3090)

Screenshot_20220911_000046

| Geomean | 1.02  |
|---------|-------|
| max     | 12.08 |
| top 1%  | 2.51  |
| top 10% | 1.01  |
| top 50% | 1.00  |
| top 90% | 0.98  |
| top 99% | 0.96  |
| min     | 0.72  |

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 1]
```diff
@@ -280,10 +292,15 @@ void maybeBuildVirtualInnerDims(
   bool split_inner_most;
   if (merged_size1 < params.tile_size1) {
     if (params.dims_merged_with_2.empty()) {
+#if 0
       // case 2
       split_inner_most = true;
       large_dim = inner_most2;
       split_factor = params.tile_size2;
+#else
+      // disabled due to indexing error
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:309]: Could you add the issue number?

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 2]
```diff
@@ -415,12 +437,6 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
 
   auto params = std::make_shared<TransposeParams>("Transpose heuristics");
 
-  // If the problem size is small use small tile sizes.
-  if (n_elems < device_multiprocessor_count * kMaxTileSize * kMaxTileSize) {
-    params->tile_size1 = 8;
-    params->tile_size2 = 8;
-  }
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:425]: Why is this removed?

**Comment by csarofeen** [comment]
Looks like when we pull this PR in we should be ready for https://github.com/csarofeen/pytorch/pull/1882/files

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:425]: Based on benchmark results, using small tile size actually hurt the performance, instead of improving it...

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 3]
```diff
@@ -212,12 +210,26 @@ class DomainMap : public pointwise_utils::DomainMap {
...
+      std::min<int64_t>(merged_size1, params.tile_size1);
+  int64_t actual_tile_size2 =
+      std::min<int64_t>(merged_size2, params.tile_size2);
+  int64_t wave_elements =
+      device_multiprocessor_count * actual_tile_size1 * actual_tile_size2;
+
+  if (wave_elements >= n_elems) {
+    // if one full wave can handle all elements, don't create virtual inner dims
+    return;
+  }
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:237]: This case should be rejected in `canScheduleRunTime`, but I think it is not a bad idea to still handle it inside the heuristics.

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 4]
```diff
@@ -334,49 +358,38 @@ void maybeBuildVirtualInnerDims(
...
-  return DomainMap::hasAtLeastTwoValidGroups(fusion);
-}
-
-std::shared_ptr<TransposeParams> getTransposeHeuristics(
-    Fusion* fusion,
-    const at::ArrayRef<c10::IValue>& runtime_inputs,
-    HeuristicSummary* data_cache) {
-  SchedulerRuntimeInfo runtime_info(fusion, runtime_inputs, true);
-  return getTransposeHeuristics(fusion, runtime_info, data_cache);
-}
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:349]: No real change, just moved below

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 5]
```diff
@@ -334,49 +358,38 @@ void maybeBuildVirtualInnerDims(
...
-    SchedulerRuntimeInfo& runtime_info,
-    HeuristicSummary* data_cache) {
-  FUSER_PERF_SCOPE("getTransposeHeuristics");
-
-  FusionGuard fg(fusion);
-
-  // Incase any buffer is of type DataType::Index
-  DataType index_type = indexModeToDtype(runtime_info.getIndexMode());
-
+HeuristicSummaryEntry<HeuristicCompileTime::DomainMap> getDomainMap(
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp]: Split-out from `getTransposeHeuristics` for code reuse. Should have no behavioral change.

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 6]
```diff
@@ -334,49 +358,38 @@ void maybeBuildVirtualInnerDims(
...
   auto domain_map_entry =
       HeuristicSummaryEntry<HeuristicCompileTime::DomainMap>(
           data_cache,
           [fusion]() { return std::make_unique<DomainMap>(fusion); });
-  const auto& domain_map = dynamic_cast<DomainMap&>(domain_map_entry.get());
+  return domain_map_entry;
+}
 
+HeuristicSummaryEntry<HeuristicCompileTime::InputsOutputsInnerDimGroups>
+getInputsOutputsGroups(HeuristicSummary* data_cache, DomainMap& domain_map) {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:378]: Split-out from `getTransposeHeuristics` for code reuse. Should have no behavioral change.

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 7]
```diff
@@ -334,49 +358,38 @@ void maybeBuildVirtualInnerDims(
...
 
   TORCH_INTERNAL_ASSERT(
       grouped_inputs_outputs.size() >= 2,
       "Can not find mismatched inner most dim, should use pointwise scheduler.");
 
+  return grouped_inputs_outputs_entry;
+}
+
+HeuristicSummaryEntry<HeuristicCompileTime::ReferenceTensors>
+getReferenceTensors(
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:395]: Split-out from `getTransposeHeuristics` for code reuse. Should have no behavioral change.

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 8]
```diff
@@ -393,13 +406,17 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
...
+  return reference_tensors_entry;
+}
 
-  const int64_t device_multiprocessor_count =
-      (int64_t)at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
-
-  auto ref_root = reference1->getMaybeRFactorDomain();
-  std::vector<int64_t> shape_in_ref1;
-  shape_in_ref1.reserve(reference1->nDims());
+std::pair<std::vector<int64_t>, int64_t> getShapeInReference(
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:418]: Split-out from `getTransposeHeuristics` for code reuse. Should have no behavioral change.

File: torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp [hunk 9]
```diff
@@ -413,30 +430,176 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
...
     n_elems *= size;
-    shape_in_ref1.push_back(size);
+    shape_in_ref.push_back(size);
   }
+  return {shape_in_ref, n_elems};
+}
 
-  auto params = std::make_shared<TransposeParams>("Transpose heuristics");
+HeuristicSummaryEntry<HeuristicCompileTime::InnerMostDimInfo>
+getInnerMostDimInfoInReference(
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp:445]: The position of innermost dim is now cached.

**Comment by zasdfgbnm** [comment]
@csarofeen I am marking this as ready, but I still need to rerun the benchmarks after fixing the effective bandwidth calculation.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 10]
```diff
@@ -937,6 +937,84 @@ class ReductionScheduler : public SchedulerEntry {
   }
 };
 
+class TransposeScheduler : public SchedulerEntry {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:940]: moved from below

**Comment by zasdfgbnm** [comment]
Should be good to go. Please check updated benchmarks https://github.com/csarofeen/pytorch/pull/1927#issue-1349786928

**Review by csarofeen — APPROVED** [review]
LGTM
